staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/validation.go

```diff
@@ -780,6 +781,23 @@ type costInfo struct {
 	// definition. A single totalCost is allocated for each validation call and passed through the stack as the
 	// custom resource definition's OpenAPIv3 schema is recursively validated.
 	TotalCost *totalCost
+	// Schema is a lazily populated structural schema for this costInfo node.
+	Schema *structuralschema.Structural
+	// DeclType is a lazily populated CEL declaration representation of Schema of this costInfo node.
+	DeclType *model.DeclType
+}
+
+// LoadSchemaIfAbsent populates the Schema and DeclType fields of costInfo if they are not already set.
+// This makes it possible for costInfos to lazily construct a structural schema and CEL DeclType only for CRD schemas that
+// have validation rules.
+func (c *costInfo) LoadSchemaIfAbsent(schema *apiextensions.JSONSchemaProps, isRoot bool) {
```

Comment:
is it valid (or if not valid, possible?) for LoadSchemaIfAbsent to be called with different values for schema and isRoot? it's a little weird to ignore params in favor of internal state on subsequent calls

Reply:
I'm going to try to eliminate this function entirely and simplify the struct lifecycle.